ampd/src/solana/verifier.rs

```diff
@@ -0,0 +1,254 @@
+use borsh::{BorshDeserialize, BorshSerialize};
+
+use base64::{self, engine::general_purpose};
+use solana_transaction_status::{
+    option_serializer::OptionSerializer, EncodedConfirmedTransactionWithStatusMeta,
+};
+use tracing::error;
+
+use crate::handlers::solana_verify_msg::Message;
+
+use super::pub_key_wrapper::PubkeyWrapper;
+
+// Gateway program logs.
+// Logged when the Gateway receives an outbound message.
+#[derive(Debug, PartialEq, BorshDeserialize, BorshSerialize, Clone)]
+#[repr(u8)]
+enum GatewayEvent {
+    CallContract {
+        // Message sender.
+        sender: PubkeyWrapper,
+        destination_chain: Vec<u8>,
+        destination_address: Vec<u8>,
+        payload: Vec<u8>,
+        payload_hash: [u8; 32],
+    },
+}
+
+impl GatewayEvent {
+    // Try to parse a [`CallContractEvent`] out of a Solana program log line.
+    fn parse_log(log: &String) -> Option<Self> {
+        let cleaned_input = log
+            .trim()
+            .trim_start_matches("Program data:")
+            .split_whitespace()
+            .flat_map(decode_base64)
+            .next()?;
+        borsh::from_slice(&cleaned_input).ok()
+    }
+}
+
+impl PartialEq<&Message> for GatewayEvent {
+    fn eq(&self, msg: &&Message) -> bool {
+        match self {
+            GatewayEvent::CallContract {
+                sender: _,
+                destination_chain,
+                destination_address,
+                payload: _,
+                payload_hash,
+            } => {
+                let event_dest_addr = String::from_utf8(destination_address.to_owned());
+                let event_dest_chain = String::from_utf8(destination_chain.to_owned());
+
+                event_dest_addr.is_ok()
+                    && event_dest_chain.is_ok()
+                    && event_dest_addr.unwrap() == msg.destination_address
+                    && msg.destination_chain == event_dest_chain.unwrap()
+                    && *payload_hash == msg.payload_hash
+            }
+        }
+    }
+}
+
+#[inline]
+fn decode_base64(input: &str) -> Option<Vec<u8>> {
+    general_purpose::STANDARD.decode(input).ok()
+}
+
+pub fn verify_message(
+    source_gateway_address: &String,
+    tx: &EncodedConfirmedTransactionWithStatusMeta,
+    message: &Message,
+) -> Vote {
+    let ui_tx = match &tx.transaction.transaction {
+        solana_transaction_status::EncodedTransaction::Json(tx) => tx,
+        _ => {
+            error!("failed to parse solana tx.");
+            return Vote::FailedOnChain;
+        }
+    };
+
+    let tx_not_found = String::from("tx not found");
+    // NOTE: first signature is always tx_id
+    let tx_id = ui_tx.signatures.first().unwrap_or(&tx_not_found);
+
+    let tx_meta = match &tx.transaction.meta {
+        Some(meta) => meta,
+        None => {
+            error!(
+                tx_id = tx_id,
+                "Theres no available tx metadata to parse log messages from."
+            );
+            return Vote::NotFound;
```

Comment:
At this stage the tx should been present (found) on the chain, right?
So maybe all errors in this function are of `FailedOnChain` type?

Reply:
I think it makes sense :+1: , fixed at 8253a8b13549213e97e6f846ffc02caa59d9f967